File: backend/src/main/java/com/interway/inventory/service/ProductService.java
package com.interway.inventory.service;

import com.interway.inventory.dto.ProductRequest;
import com.interway.inventory.dto.ProductResponse;
import com.interway.inventory.exception.ProductNotFoundException;
import com.interway.inventory.model.Product;
import com.interway.inventory.repository.ProductRepository;
import com.interway.inventory.repository.ProductSpecifications;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.multipart.MultipartFile;


import java.math.BigDecimal;
import java.util.List;

@Service
@Transactional
public class ProductService {
    private final ProductRepository repository;
    private final FileStorageService fileStorageService;

    public ProductService(ProductRepository repository, FileStorageService fileStorageService) {
        this.repository = repository;
        this.fileStorageService = fileStorageService;
    }

    @Transactional(readOnly = true)
    public Page<ProductResponse> findAll(String search, String category,
                                         BigDecimal minPrice, BigDecimal maxPrice,
                                         Pageable pageable) {
        Specification<Product> spec = Specification.unrestricted();

        if (search != null && !search.isBlank()) {
            spec = spec.and(ProductSpecifications.nameContains(search));
        }
        if (category != null && !category.isBlank()) {
            spec = spec.and(ProductSpecifications.categoryEquals(category));
        }
        if (minPrice != null) {
            spec = spec.and(ProductSpecifications.priceAtLeast(minPrice));
        }
        if (maxPrice != null) {
            spec = spec.and(ProductSpecifications.priceAtMost(maxPrice));
        }

        return repository.findAll(spec, pageable).map(this::toResponse);
    }

    @Transactional(readOnly = true)
    public ProductResponse findById(Long id){
        Product product = repository.findById(id).orElseThrow(() -> new ProductNotFoundException(id));
        return toResponse(product);
    }

    public ProductResponse create(ProductRequest request){
        Product product = new Product();
        applyRequest(product, request);
        return toResponse(repository.save(product));
    }

    public ProductResponse update(Long id, ProductRequest request){
        Product product = repository.findById(id).orElseThrow(() -> new ProductNotFoundException(id));
        applyRequest(product, request);
        return toResponse(repository.save(product));
    }

    public void delete(Long id) {
        Product product = repository.findById(id)
                .orElseThrow(() -> new ProductNotFoundException(id));
        String image = product.getImageUrl();
        repository.deleteById(id);
        fileStorageService.delete(image);
    }

    private void applyRequest(Product product, ProductRequest request) {
        product.setName(request.getName());
        product.setDescription(normalize(request.getDescription()));
        product.setPrice(request.getPrice());
        product.setQuantityInStock(request.getQuantityInStock());
        product.setCategory(normalize(request.getCategory()));
    }

    private String normalize(String value) {
        return (value == null || value.isBlank() ? null : value.trim());
    }

    private ProductResponse toResponse(Product product) {
        ProductResponse response = new ProductResponse();
        response.setId(product.getId());
        response.setName(product.getName());
        response.setDescription(product.getDescription());
        response.setPrice(product.getPrice());
        response.setQuantityInStock(product.getQuantityInStock());
        response.setCategory(product.getCategory());
        response.setImageUrl(product.getImageUrl());
        response.setCreatedAt(product.getCreatedAt());
        response.setUpdatedAt(product.getUpdatedAt());
        return response;
    }

    @Transactional(readOnly = true)
    public List<String> findAllCategories(){
        return repository.findDistinctCategories();
    }

    public ProductResponse uploadImage(Long id, MultipartFile file) {
        Product product = repository.findById(id)
                .orElseThrow(() -> new ProductNotFoundException(id));

        String oldImage = product.getImageUrl();
        String filename = fileStorageService.store(file);
        product.setImageUrl(filename);

        Product saved = repository.save(product);
        fileStorageService.delete(oldImage);

        return toResponse(saved);
    }

    public ProductResponse deleteImage(Long id) {
        Product product = repository.findById(id)
                .orElseThrow(() -> new ProductNotFoundException(id));

        String filename = product.getImageUrl();
        product.setImageUrl(null);
        Product saved = repository.save(product);
        fileStorageService.delete(filename);

        return toResponse(saved);
    }

}

File: frontend/src/app/services/product.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Product, Page } from '../models/product.model';

@Injectable({ providedIn: 'root' })
export class ProductService {
  private readonly baseUrl = 'http://localhost:8080/products';

  constructor(private http: HttpClient) {}

  getAll(
    search?: string,
    category?: string,
    minPrice?: number,
    maxPrice?: number,
    page = 0,
    size = 10,
  ): Observable<Page<Product>> {
    let params = new HttpParams().set('page', page).set('size', size);

    if (search) params = params.set('search', search);
    if (category) params = params.set('category', category);
    if (minPrice != null) params = params.set('minPrice', minPrice);
    if (maxPrice != null) params = params.set('maxPrice', maxPrice);

    return this.http.get<Page<Product>>(this.baseUrl, { params });
  }

  getById(id: number): Observable<Product> {
    return this.http.get<Product>(`${this.baseUrl}/${id}`);
  }

  create(product: Product): Observable<Product> {
    return this.http.post<Product>(this.baseUrl, product);
  }

  update(id: number, product: Product): Observable<Product> {
    return this.http.put<Product>(`${this.baseUrl}/${id}`, product);
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }

  getCategories(): Observable<string[]> {
    return this.http.get<string[]>(`${this.baseUrl}/categories`);
  }

  uploadImage(id: number, file: File): Observable<Product> {
    const formData = new FormData();
    formData.append('file', file);
    return this.http.post<Product>(`${this.baseUrl}/${id}/image`, formData);
  }

  getImageUrl(filename?: string): string | null {
    return filename ? `http://localhost:8080/uploads/${filename}` : null;
  }

  deleteImage(id: number): Observable<Product> {
    return this.http.delete<Product>(`${this.baseUrl}/${id}/image`);
  }
}
